Replace `nmtoken` with `unquoted` (#364)
unicode-org/message-format-wg: spec/syntax.md

but an _expression_ must not be completely empty.

```abnf
expression = "{" [s] (((literal / argtoken / variable) [s annotation]) / annotation) [s] "}"
annotation = function *(s option)
option = name [s] "=" [s] (literal / nmtoken / variable)
```

Expression examples:

```

> {1.23}

Reviewer: we shouldn't remove the quoted examples, since they are still valid. We can augment the examples with unquoted ones. It would be useful to add or change the quoted ones to be something where the quotes add value, e.g. perhaps more like:

```
{1.23}
{|1.23|}
{1.23 :number maxFractionDigits=1}
{|-1.23| :number maxFractionDigits=2}
{|My Brand Name| :linkify href=|foobar.com|}
```

Author: I added a couple such examples.

          / %xE000-10FFFF
```

### Literals

_Literal_ is used for matching variants and providing input to _expressions_.
_Quoted_ literals may include content with any Unicode code point,
except for surrogate code points U+D800 through U+DFFF.
The characters `\` and `|` MUST be escaped as `\\` and `\|`.
_Unquoted_ literals have a much more restricted range that
is intentionally close to the XML's [Nmtoken](https://www.w3.org/TR/xml/#NT-Nmtoken),
with the restriction that it MUST NOT start with `-` or `:`,

> as those would conflict with _function_ start characters.

Reviewer 2: These two lines don't feel _normative_ the way the rest of this passage does. Perhaps remove them?

Reviewer: This isn't quite how I would approach this. I think it needs more explanation about literals in general and then about the unquoted ones. The relationship to Nmtoken (which we broke on purpose) isn't that relevant any more. Perhaps:

```suggestion
_Literal_ values are used to pass data to various parts of a `message`:
* As the value of a `key` in a `when` statement
* As the `argument` in an `expression`
* As the `value` in an `option`

A `Literal` is a sequence of _Unicode code points_ and can include any Unicode character. Surrogate code points are not allowed. 

The characters `\\` U+005C REVERSE SOLIDUS and `|` U+007C VERTICAL BAR **_must_** be escaped (as `\\` and `\|` respectively) when they appear in the value of a `Literal`.

Spaces are significant in a `Literal`.

A `Quoted` literal is surrounded by `|` characters. 

A `Literal` can be `Unquoted` when its content matches that production. The content restrictions for `Unquoted` follow best practices for the use of Unicode in formal grammars and are intentionally similar to, for example, XML's [Nmtoken](https://www.w3.org/TR/xml/#NT-Nmtoken).
```

Reviewer 3: > +A `Literal` is a sequence of _Unicode code points_ and can include any Unicode character. Surrogate code points are not allowed. 

Should be:

A `Literal` is a sequence of _Unicode code points_, and can contain any Unicode code points except for surrogate code points and non-character code points.

Reason: "Unicode character" would mean "assigned Unicode character", which is unnecessarily fragile across versions.

Reviewer: Actually, non-characters (U+FFFF for example) are not excluded. Only surrogate code points are. This is consistent with e.g. DOMString and USVString.

Reviewer 3: That's fine

All code points are preserved.

```abnf
literal = quoted / unquoted

quoted         = "|" *(quoted-char / quoted-escape) "|"
quoted-char    = %x0-5B         ; omit \
               / %x5D-7B        ; omit |
               / %x7D-D7FF      ; omit surrogates
               / %xE000-10FFFF

unquoted       = unquoted-start *name-char
unquoted-start = name-start / DIGIT / "."
               / %xB7 / %x300-36F / %x203F-2040
```

### Names

The _name_ token is used for variable names (prefixed with `$`),
function names (prefixed with `:`, `+` or `-`),
as well as option names.
It is based on XML's [Name](https://www.w3.org/TR/xml/#NT-Name),

> with the restriction that it MUST NOT start with an ASCII digit and certain basic combining characters.

Reviewer 2: The phrasing here suggested that XML's `Name` allows digits as the first character.

Author: Good point. The only actual divergence is the colon.

```suggestion
with the restriction that it MUST NOT start with `:`,
as that would conflict with _function_ start characters.
```

The _nmtoken_ token doesn't have _name_'s restriction on the first character
and is used as variant keys and option values.

The _argtoken_ is used for expression arguments,

> and matches _nmtoken_ except for not allowing `-` as a first character.

Reviewer: As above, perhaps add text about why `argtoken` exists, e.g.:

> `argtoken` exists as a convenience, since it allows some literal values to be unquoted.

Author: Done.

_Note:_ The Name and Nmtoken symbols are intentionally defined to be
the same as XML's [Name](https://www.w3.org/TR/xml/#NT-Name) and [Nmtoken](https://www.w3.org/TR/xml/#NT-Nmtokens)

> in order to increase the interoperability with data defined in XML.

Reviewer: code quote `Name` and `Nmtoken` and match the casing in the ABNF. I wouldn't call them "symbols". I would probably call them productions (since that's the ABNF terminology) or just "nothing", i.e. "_Note:_ `name` and `nmtoken` are intentionally defined..."

I would also be more direct here, changing "... in order to increase the interoperability with data defined in XML" to something more like:

> ... in order to ensure current and future compatibility with LDML (UAX35) and it's various names and values, since these are defined using XML.

Author: Should these changes be done in a separate non-normative editorial PR?

The _nmtoken_ token doesn't have _name_'s restriction on the first character
and is used as variant keys and option values.

The _argtoken_ is used for expression arguments,
and matches _nmtoken_ except for not allowing `-` or `:` as a first character.

> It allows some literal argument values to be unquoted.

Reviewer: This is the part of this PR that keeps bringing me back for another look. I don't know what an "expression argument" is, especially given that other tokens found inside an expression can be considered arguments (notably, `options` can be considered arguments to a `function` or a `reserved`).

Since the purpose of `argtoken` is to allow unquoted literals, I think I might prefer if we dealt with them by recasting `literal` and naming `argtoken` as `unquoted`, e.g.:

```abnf
; literal data values; some values, depending on content, can appear
; without quoting | delimiters
literal  = unquoted / ("|" *(literal-char / literal-escape) "|")
unquoted = arg-start *name-char
```

Otherwise I'd want to rewrite aspects of this text to clarify "expression argument"

Author: I've now renamed the added tokens as `unquoted` and `unquoted-start`.

> The _unquoted_ token is used for expression arguments,
> and matches _nmtoken_ except for not allowing `-` or `:` as a first character.
> It allows some literal argument values to be unquoted.

Author: ```suggestion
The _unquoted_ token allows some literal values to appear unquoted in an _expression_.
It matches _nmtoken_ except for not allowing `-` or `:` as a first character.
```
